File: golangLeetcode/283.移动零.go
/*
 * @lc app=leetcode.cn id=283 lang=golang
 *
 * [283] 移动零
 */

// @lc code=start
func moveZeroes(nums []int) {
	if len(nums) < 2 {
		return
	}
	// method 1
	// zIndex := -1
	// for i := 0; i < len(nums); {
	// 	if zIndex != -1 {
	// 		if nums[i] != 0 {
	// 			nums[i], nums[zIndex] = nums[zIndex], nums[i]
	// 			i = zIndex
	// 			zIndex = -1
	// 		} else {
	// 			i++
	// 		}
	// 	} else {
	// 		if nums[i] == 0 {
	// 			zIndex = i
	// 		}
	// 		i++
	// 	}
	// }

	// methd 2
	for i, j := 0, 0; i < len(nums) && j < len(nums); {
		if nums[i] == 0 && nums[j] != 0 {
			nums[i], nums[j] = nums[j], nums[i]
			j++
		} else if nums[i] == 0 && nums[j] == 0 {
			j++
		} else if nums[i] != 0 && nums[j] == 0 {
			i++
		} else {
			i++
			if i > j {
				j++
			}
		}
	}

}

// @lc code=end

File: LoveSwiftLeetcode/283.移动零.swift
/*
 * @lc app=leetcode.cn id=283 lang=swift
 *
 * [283] 移动零
 */

// @lc code=start
class Solution {
    func removeElement(_ nums: inout [Int], _ val: Int) -> Int {
        var fast = 0, slow = 0
        while fast < nums.count {
            if nums[fast] != val {
                nums[slow] = nums[fast]
                slow += 1
            }
            fast += 1
        }
        return slow
    }
    func moveZeroes(_ nums: inout [Int]) {
        let index = removeElement(&nums, 0)
        for i in index..<nums.count {
            nums[i] = 0
        }
    }
}
// @lc code=end

